### PR TITLE
Rescue from ActiveJob::DeserializationError during derivative creation.

### PR DESCRIPTION
**Pull request by eddierubeiz**
Ref https://github.com/sciencehistory/scihist_digicoll/issues/577 .
Includes a test. Thanks, @jrochkind, for excellent pointers on how to deal with this.

### CODE REVIEW

File: app/jobs/kithe/create_derivatives_job.rb [hunk 1]
```diff
@@ -1,7 +1,12 @@
 module Kithe
   class CreateDerivativesJob < Job
     def perform(asset, mark_created: nil)
-      asset.create_derivatives(mark_created: mark_created)
+      begin
+        asset.create_derivatives(mark_created: mark_created)
+      rescue ActiveJob::DeserializationError
+        Rails.logger.error("Unable to derivatives for asset with ID #{asset.id}, as it was deleted first.")
```
Comment (jrochkind) [inline app/jobs/kithe/create_derivatives_job.rb:7]: Maybe some weirder error happened than the asset being deleted, so message maybe should say "as the asset model was unavailable". 

I'm not totally sure that `asset.id` will be available in this condition, did you verify that?

There may be more info available in the `ActiveJob::DeserializationError` itself, if you change to `ActiveJob::DeserializationError => e`, you can get the exception in local variable `e`, and maybe actually print out it's representation in the error message too?

    Rails.logger.error("Unable to derivatives for asset with ID #{asset.id}, as it was not available: #{e}")

But make sure to actually repro at least in dev and see what the error looks like, and that it looks useful. To just record verification that we checked this, please put a copy paste of what final error looks like in this ticket -- or better yet, find a way to test it somehow! 

Perhaps make an asset, delete it, then manually trigger a CreateDerivativesJob for it. There is some hacky way to test Rails logger output in a test, but I'd have to google it.

**Comment by eddierubeiz** [comment]
Stack trace from the bug as I reproduced it in staging:

```Class Kithe::CreateDerivativesJob (via ActiveJob)
Arguments
job_class: Kithe::CreateDerivativesJob
job_id: 842e61fe-bd0b-4925-823d-d6dfc6cc72ec
provider_job_id: 
queue_name: default
priority: 
arguments:
- _aj_globalid: gid://scihist-digicoll/Asset/4886ae27-1c10-45ca-baf0-bb6593364aa2
executions: 0
exception_executions: {}
locale: en
timezone: US/Eastern
enqueued_at: '2020-04-23T17:57:26Z'

Exception ActiveJob::DeserializationError

Error
Error while trying to deserialize arguments: Couldn't find Asset with 'id'=4886ae27-1c10-45ca-baf0-bb6593364aa2 [WHERE "kithe_models"."type" IN ($1, $2)]
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/relation/finder_methods.rb:339:in `raise_record_not_found_exception!'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/relation/finder_methods.rb:452:in `find_one'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/relation/finder_methods.rb:434:in `find_with_ids'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/relation/finder_methods.rb:69:in `find'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/querying.rb:21:in `find'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/core.rb:160:in `find'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/globalid-0.4.2/lib/global_id/locator.rb:130:in `locate'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/globalid-0.4.2/lib/global_id/locator.rb:155:in `block in locate'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/globalid-0.4.2/lib/global_id/locator.rb:165:in `block in unscoped'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/scoping/default.rb:34:in `block in unscoped'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/relation.rb:398:in `block in scoping'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/relation.rb:773:in `_scoping'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/relation.rb:398:in `scoping'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activerecord-6.0.2.2/lib/active_record/scoping/default.rb:34:in `unscoped'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/globalid-0.4.2/lib/global_id/locator.rb:165:in `unscoped'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/globalid-0.4.2/lib/global_id/locator.rb:155:in `locate'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/globalid-0.4.2/lib/global_id/locator.rb:17:in `locate'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/arguments.rb:117:in `deserialize_global_id'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/arguments.rb:101:in `deserialize_argument'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/arguments.rb:42:in `block in deserialize'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/arguments.rb:42:in `map'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/arguments.rb:42:in `deserialize'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/core.rb:168:in `deserialize_arguments'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/core.rb:158:in `deserialize_arguments_if_needed'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/execution.rb:37:in `perform_now'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/execution.rb:24:in `block in execute'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activesupport-6.0.2.2/lib/active_support/callbacks.rb:112:in `block in run_callbacks'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/railtie.rb:43:in `block (4 levels) in <class:Railtie>'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activesupport-6.0.2.2/lib/active_support/execution_wrapper.rb:88:in `wrap'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activesupport-6.0.2.2/lib/active_support/reloader.rb:72:in `block in wrap'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activesupport-6.0.2.2/lib/active_support/execution_wrapper.rb:88:in `wrap'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activesupport-6.0.2.2/lib/active_support/reloader.rb:71:in `wrap'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/railtie.rb:42:in `block (3 levels) in <class:Railtie>'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activesupport-6.0.2.2/lib/active_support/callbacks.rb:121:in `instance_exec'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activesupport-6.0.2.2/lib/active_support/callbacks.rb:121:in `block in run_callbacks'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activesupport-6.0.2.2/lib/active_support/callbacks.rb:139:in `run_callbacks'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/execution.rb:22:in `execute'
/opt/scihist_digicoll/shared/bundle/ruby/2.6.0/gems/activejob-6.0.2.2/lib/active_job/queue_adapters/resque_adapter.rb:47:in `perform'
```

File: app/jobs/kithe/create_derivatives_job.rb [hunk 2]
```diff
@@ -1,7 +1,13 @@
...
   class CreateDerivativesJob < Job
     def perform(asset, mark_created: nil)
-      asset.create_derivatives(mark_created: mark_created)
+        asset.create_derivatives(mark_created: mark_created)
+    end
+    # This error typically occurs when several large assets, whose derivatives
+    # take a long time to generate, are deleted immediately after ingest.
+    rescue_from(ActiveJob::DeserializationError) do |exception|
+      Rails.logger.error("Unable to derivatives for this asset, as it was unavailable.")
+      Rails.logger.error("Details: #{exception.inspect}")
```
Comment (jrochkind) [inline app/jobs/kithe/create_derivatives_job.rb:10]: Two separate `Rails.logger.error` lines risks that some other log entry could end up being logged in between them. Better to do one single `Rails.logger.error`. If you want multiple lines, you can still put a `"\n"` in there. But i'm not sure it needs to be multiple lines, what about just `Unable to delete deravatives for this asset, as it was unavailable: #{exception}"`? (Not sure what difference adding explicit `.inspect` makes if any). (Note also missing verb). 

Thanks for pasting the stack trace from the uncaught exception. But what I really want to see is what the *log* looks like from *this* code, what it actually looks like when you log it.

**Comment by eddierubeiz** [comment]
Took a while, but I was able to reproduce in staging: with the new code, the error is indeed caught and logged to `log/production.log`:

```E, [2020-04-28T12:25:20.135421 #16278] ERROR -- : Kithe::CreateDerivativesJob: Unable to create derivatives for this asset, as it was unavailable. Details: Error while trying to deserialize arguments: Couldn't find Asset with 'id'=a8c39ccf-9ab6-4e79-9ed1-f7873ec0231d [WHERE "kithe_models"."type" IN ($1, $2)]```

**Pull request merged**